### ISSUE TITLE
Crash with parallel=True for array shapes computed via tuple concatenation

### ISSUE THREAD

**Issue by eric-wieser** [numba/numba#5383]
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG). (`numba.__version__ == '0.48.0'`)
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Offending code:
```python
import numba
import numpy as np

@numba.njit(parallel=True, nogil=True)
def _numba_construct_tables(gradeList):
    n = len(gradeList)
    array_shape = n, n
    indices = np.zeros((1,) + array_shape, dtype=np.uint64)
    k_list = indices[0, :]

    for i, grade_list_i in enumerate(gradeList):
        k_list[i, i] = i

_numba_construct_tables(np.array([1, 1]),)
```

<details>

<summary>
Traceback:
</summary>

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-824e0b6c5461> in <module>
     13             k_list[i, j] = i + j
     14
---> 15 _numba_construct_tables(np.array([1, 1]),)

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\dispatcher.py in _compile_for_args(self, *args, **kws)
    418                     e.patch_message('\n'.join((str(e).rstrip(), help_msg)))
    419             # ignore the FULL_TRACEBACKS config, this needs reporting!
--> 420             raise e
    421
    422     def inspect_llvm(self, signature=None):

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\dispatcher.py in _compile_for_args(self, *args, **kws)
    351                 argtypes.append(self.typeof_pyval(a))
    352         try:
--> 353             return self.compile(tuple(argtypes))
    354         except errors.ForceLiteralArg as e:
    355             # Received request for compiler re-entry with the list of arguments

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\compiler_lock.py in _acquire_compile_lock(*args, **kwargs)
     30         def _acquire_compile_lock(*args, **kwargs):
     31             with self:
---> 32                 return func(*args, **kwargs)
     33         return _acquire_compile_lock
     34

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\dispatcher.py in compile(self, sig)
    766             self._cache_misses[sig] += 1
    767             try:
--> 768                 cres = self._compiler.compile(args, return_type)
    769             except errors.ForceLiteralArg as e:
    770                 def folded(args, kws):

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\dispatcher.py in compile(self, args, return_type)
     75
     76     def compile(self, args, return_type):
---> 77         status, retval = self._compile_cached(args, return_type)
     78         if status:
     79             return retval

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\dispatcher.py in _compile_cached(self, args, return_type)
     89
     90         try:
---> 91             retval = self._compile_core(args, return_type)
     92         except errors.TypingError as e:
     93             self._failed_cache[key] = e

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\dispatcher.py in _compile_core(self, args, return_type)
    102
    103         impl = self._get_implementation(args, {})
--> 104         cres = compiler.compile_extra(self.targetdescr.typing_context,
    105                                       self.targetdescr.target_context,
    106                                       impl,

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\compiler.py in compile_extra(typingctx, targetctx, func, args, return_type, flags, locals, library, pipeline_class)
    549     pipeline = pipeline_class(typingctx, targetctx, library,
    550                               args, return_type, flags, locals)
--> 551     return pipeline.compile_extra(func)
    552
    553

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\compiler.py in compile_extra(self, func)
    329         self.state.lifted = ()
    330         self.state.lifted_from = None
--> 331         return self._compile_bytecode()
    332
    333     def compile_ir(self, func_ir, lifted=(), lifted_from=None):

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\compiler.py in _compile_bytecode(self)
    391         """
    392         assert self.state.func_ir is None
--> 393         return self._compile_core()
    394
    395     def _compile_ir(self):

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\compiler.py in _compile_core(self)
    371                 self.state.status.fail_reason = e
    372                 if is_final_pipeline:
--> 373                     raise e
    374         else:
    375             raise CompilerError("All available pipelines exhausted")

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\compiler.py in _compile_core(self)
    362             res = None
    363             try:
--> 364                 pm.run(self.state)
    365                 if self.state.cr is not None:
    366                     break

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\compiler_machinery.py in run(self, state)
    345                     (self.pipeline_name, pass_desc)
    346                 patched_exception = self._patch_error(msg, e)
--> 347                 raise patched_exception
    348
    349     def dependency_analysis(self):

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\compiler_machinery.py in run(self, state)
    336                 pass_inst = _pass_registry.get(pss).pass_inst
    337                 if isinstance(pass_inst, CompilerPass):
--> 338                     self._runPass(idx, pass_inst, state)
    339                 else:
    340                     raise BaseException("Legacy pass in use")

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\compiler_lock.py in _acquire_compile_lock(*args, **kwargs)
     30         def _acquire_compile_lock(*args, **kwargs):
     31             with self:
---> 32                 return func(*args, **kwargs)
     33         return _acquire_compile_lock
     34

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\compiler_machinery.py in _runPass(self, index, pss, internal_state)
    300             mutated |= check(pss.run_initialization, internal_state)
    301         with SimpleTimer() as pass_time:
--> 302             mutated |= check(pss.run_pass, internal_state)
    303         with SimpleTimer() as finalize_time:
    304             mutated |= check(pss.run_finalizer, internal_state)

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\compiler_machinery.py in check(func, compiler_state)
    273
    274         def check(func, compiler_state):
--> 275             mangled = func(compiler_state)
    276             if mangled not in (True, False):
    277                 msg = ("CompilerPass implementations should return True/False. "

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\typed_passes.py in run_pass(self, state)
    267                                          state.flags,
    268                                          state.parfor_diagnostics)
--> 269         parfor_pass.run()
    270
    271         # check the parfor pass worked and warn if it didn't

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\parfor.py in run(self)
   1507             self._convert_setitem(self.func_ir.blocks)
   1508         if self.options.numpy:
-> 1509             self._convert_numpy(self.func_ir.blocks)
   1510         if self.options.reduction:
   1511             self._convert_reduce(self.func_ir.blocks)

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\parfor.py in _convert_numpy(self, blocks)
   1637                         # only translate C order since we can't allocate F
   1638                         if guard(self._is_supported_npycall, expr):
-> 1639                             instr = self._numpy_to_parfor(equiv_set, lhs, expr)
   1640                             if isinstance(instr, tuple):
   1641                                 pre_stmts, instr = instr

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\parfor.py in _numpy_to_parfor(self, equiv_set, lhs, expr)
   2363         kws = dict(expr.kws)
   2364         if call_name in ['zeros', 'ones'] or mod_name == 'numpy.random':
-> 2365             return self._numpy_map_to_parfor(equiv_set, call_name, lhs, args, kws, expr)
   2366         # return error if we couldn't handle it (avoid rewrite infinite loop)
   2367         raise NotImplementedError("parfor translation failed for ", expr)

~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\parfor.py in _numpy_map_to_parfor(self, equiv_set, call_name, lhs, args, kws, expr)
   2377         # generate loopnests and size variables from lhs correlations
   2378         size_vars = equiv_set.get_shape(lhs)
-> 2379         index_vars, loopnests = self._mk_parfor_loops(size_vars, scope, loc)
   2380
   2381         # generate init block and body

```

</details>

```
~\AppData\Local\Programs\Python\Python38\lib\site-packages\numba\parfor.py in _mk_parfor_loops(self, size_vars, scope, loc)
   2181         loopnests = []
   2182         index_vars = []
-> 2183         for size_var in size_vars:
   2184             index_var = ir.Var(scope, mk_unique_var("parfor_index"), loc)
   2185             index_vars.append(index_var)

TypeError: Failed in nopython mode pipeline (step: convert to parfors)
'NoneType' object is not iterable
```

---

Avoiding the tuple concatenation works fine
```python
import numba
import numpy as np

@numba.njit(parallel=True, nogil=True)
def _numba_construct_tables(gradeList):
    n = len(gradeList)
    indices = np.zeros((1, n, n), dtype=np.uint64)
    k_list = indices[0, :]

    for i, grade_list_i in enumerate(gradeList):
        k_list[i, i] = i

_numba_construct_tables(np.array([1, 1]),)
```
as does not using parallel mode:
```python
import numba
import numpy as np

@numba.njit(parallel=False, nogil=True)
def _numba_construct_tables(gradeList):
    n = len(gradeList)
    array_shape = n, n
    indices = np.zeros((1,) + array_shape, dtype=np.uint64)
    k_list = indices[0, :]

    for i, grade_list_i in enumerate(gradeList):
        k_list[i, i] = i

_numba_construct_tables(np.array([1, 1]),)
```

**Comment by stuartarchibald** [comment]
Thanks for the report. Think this is a bug in creating a parfor from a compound expression.

CC @DrTodd13

**Comment by stuartarchibald** [comment]
Closing as fixed by #6224, thanks!
